[WIP] Kotlin samples for the logging user manual chapter

eriwen opened the pull request:
This change introduces Kotlin DSL samples for the logging chapter.

It leverages a new `include::sample` asciidoctor macro I wrote (sources not open yet, but internal link is [here](https://github.com/gradle/dotorg-docs/tree/master/docs-asciidoctor-extensions)).

Items for discussion that we want to definitively decide before we write a whole lot of Kotlin DSL samples:

 * We are completely cool using this task creation syntax for Kotlin examples. 
   * I think the Groovy and Kotlin samples should be as close as possible. This implies that we should _not_ use lazy configuration Kotlin APIs _unless_ we also use them for Groovy DSL.
 * Should we change anything _now_ about how we declare samples and expected output? With Exemplar we now have the ability to extract samples and output from the docs and test it. 
   * I think @adammurdoch and I agreed that the sample _sources should remain external_ but perhaps the _expected output should be embedded_. Do we want to do this now?

NOTE: Please ignore changes to `head.html`. I will not merge these but instead fix `https://docs.gradle.org/build-tool.js` to work even if the docs are served locally. 

![screenflow](https://user-images.githubusercontent.com/51534/43930721-d78ea038-9bef-11e8-9edf-10bdf989b0fc.gif)

big-guy commented on subprojects/docs/src/samples/userguide/tutorial/loggingKts/build.gradle.kts:
```diff
@@ -0,0 +1,39 @@
+// end::capture-stdout[]
+
+// tag::task-capture-stdout[]
+tasks.create("logInfo") {
```
I think this should be something like:

```
tasks {
    val logInfo by registering {
        logging.captureStandardOutput(LogLevel.INFO)
        doFirst {
           println("A task message which is logged at INFO level")
        }
    }
}
```

eriwen replied:
I don't think we want to use delegated properties syntax here, but we might want to use the `tasks {}` block. WDYT @eskatos?

eskatos replied:
We should try make the Groovy/Kotlin samples as similar as possible.

The behavior *must* be the same, so no lazy configuration api usage when the Groovy sample is not using lazy configuration.

FWIW, the strict equivalent of the Groovy:
```groovy
task logInfo {
    logging.captureStandardOutput LogLevel.INFO
    doFirst {
        println 'A task message which is logged at INFO level'
    }
}
```

would be the Kotlin:
```kotlin
task("logInfo") {
    logging.captureStandardOutput(LogLevel.INFO)
    doFirst {
        println("A task message which is logged at INFO level")
    }
}
```

eskatos replied:
Delegated properties are useful when one need a reference to the created/configured task for later use in the same script.

adammurdoch commented on subprojects/docs/src/docs/userguide/logging.adoc:
```diff
@@ -104,12 +104,7 @@ You can also hook into Gradle's logging system from within other classes used in
-include::{samplesPath}/userguide/tutorial/logging/build.gradle[tag=use-slf4j]
-----
-
+include::sample[dir="userguide/tutorial/logging",files="build.gradle[tags=use-slf4j]"]
```
it would be good to also update the README to describe the syntax that should be used in the user guide as part of this PR (to make it easier to understand what the proposed syntax is).

adammurdoch commented on subprojects/docs/src/docs/userguide/logging.adoc:
```diff
@@ -143,19 +136,23 @@ Gradle also provides integration with the Java Util Logging, Jakarta Commons Log
+[.multi-language-text.lang-kotlin]
+----
+> gradle -I customLoggerKts.init.gradle.kts build
+include::{samplesPath}/userguide/initScripts/customLoggerKts/customLoggerKts.out[]
```
In general, we want all samples to behave the same way regardless of whether groovy dsl or kotlin dsl is being used. Which means the expected output should be the _same_ for each variant of the sample, which implies there should be only one expected output for a sample that is used for both DSL language variants.

adammurdoch commented on subprojects/docs/src/samples/userguide/tutorial/loggingKts/settings.gradle.kts:
```diff
@@ -0,0 +1 @@
+rootProject.name = "loggingKts"
```
I think it would be better if each variant of a sample had the same Gradle structure, e.g. same project names.

eskatos replied:
👍 to what Adam said above.

Sample directories named something like `logging-groovy` and `logging-kotlin` would make them more obvious to users browsing samples than the cryptic `Kts` suffix.

eriwen replied:
okay

big-guy commented on subprojects/docs/src/docs/userguide/logging.adoc:
```diff
@@ -118,23 +113,21 @@ Internally, Gradle uses Ant and Ivy. Both have their own logging system. Gradle
-----
-include::{samplesPath}/userguide/tutorial/logging/build.gradle[tag=capture-stdout]
-----
+include::sample[dir="userguide/tutorial/loggingKts",files="build.gradle.kts[tags=capture-stdout]"]
```
can we do something with this extension so that we enforce that kotlin and groovy samples are named consistently?  I think it would also be OK if we arranged the samples more in a hierarchy (`userguide/tutorial/logging/kotlin` and `userguide/tutorial/logging/groovy`).

JLLeitschuh commented on subprojects/docs/src/samples/userguide/initScripts/customLoggerKts/build.gradle.kts:
```diff
@@ -0,0 +1,20 @@
+}
+tasks.create("build") {
+    dependsOn("test")
+}
```
Do you want to use the new deferred task configuration API?

eskatos replied:
The side by side Groovy and Kotlin samples should behave the same way. Moving to the configuration avoidance APIs should be done separately and uniformly.

eriwen commented:
@adammurdoch @eskatos @big-guy I think I've addressed all feedback, save converting to use lazy configuration APIs because we agreed not to do this yet.

Here's what I changed: 
 * Groovy and Kotlin sample projects now have their own subdirectories, but there is only 1 `.sample.conf` and output file for each, ensuring they are executed and verified the same.
 * Used the `include::sample` everywhere in logging.adoc. 
 * Fixed up a couple rough edges in the Kotlin samples

Mergeable now?

eskatos commented on subprojects/docs/src/samples/userguide/initScripts/customLogger/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,20 @@
+tasks.create("compile") {
```
:question: didn't we agree to use `task("compile") { ... }` here to let the Kotlin script do exactly and look the same as the Groovy script?

eskatos commented on subprojects/docs/src/samples/userguide/initScripts/customLogger/kotlin/customLogger.init.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+
+    override fun buildFinished(result: BuildResult) {
+        println("build completed")
+        result.failure?.printStackTrace()
```
❓ Couldn't the corresponding Groovy sample use the same syntax here: `result.failure?.printStackTrace()`?

This would have the benefit to let them have the same number of lines, preventing a layout change when switching DSL languages.

BTW, the `public` modifier on overridden methods in the Groovy sample are unnecessary and could be omitted.

eriwen replied:
Yeah, will change. Note that the page should not jump when changing DSLs anymore. If you still see this, please send me details so I can fix it.

eskatos replied:
It's not jumping anymore since your fix, but the layout around change, nothing we can do about it except make samples the same length when it makes sense.

eskatos commented on subprojects/docs/src/samples/userguide/tutorial/logging/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,39 @@
+// tag::use-slf4j[]
+import org.slf4j.Logger
```
:x: unused import

eskatos commented on subprojects/docs/src/samples/userguide/tutorial/logging/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,39 @@
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+val slf4jLogger = LoggerFactory.getLogger("some-logger")
```
:o: nit pic, the corresponding Groovy sample uses the following declaration: `Logger slf4jLogger = ...` which makes the `org.slf4j.Logger` import necessary. It could be changed to `def slf4jLogger = ...` to make it looks more similar to the Kotlin sample.

eskatos commented on subprojects/docs/README.md:
```diff
@@ -64,59 +64,65 @@ You will find these references useful when authoring AsciiDoc:
+----
+[.multi-language-text.lang-kotlin]
+----
+$ gradle -I customLoggerKts.init.gradle.kts build
```
:x: should be `customLogger.init.gradle.kts`

eskatos commented on subprojects/docs/src/docs/userguide/logging.adoc:
```diff
@@ -143,18 +130,21 @@ Gradle also provides integration with the Java Util Logging, Jakarta Commons Log
 ----
-> gradle -I init.gradle build
-include::{samplesPath}/userguide/initScripts/customLogger/custom_logging_ui.out[]
+$ gradle -I customLoggerKts.init.gradle.kts build
```
:x: should be `customLogger.init.gradle.kts`

eskatos commented on subprojects/docs/src/samples/userguide/initScripts/customLogger/kotlin/settings.gradle.kts:
```diff
@@ -0,0 +1 @@
+rootProject.name = "customLoggerKts"
```
❓ `rootProject.name = "customLogger"`

eskatos commented on subprojects/docs/README.md:
```diff
@@ -64,59 +64,65 @@ You will find these references useful when authoring AsciiDoc:
+[.multi-language-text.lang-groovy]
 ----
-include::{samples-path}/foo/build.gradle[tag=something]
+$ gradle -I init.gradle build
```
❓ the Groovy init script could be renamed `customLogger.init.gradle` or `customLogger.gradle`

eriwen commented:
Squashed/Merged with 65e902a
